tools/wake-format/emitter.cpp

```diff
@@ -1030,7 +1030,42 @@ wcl::doc Emitter::walk_lambda(ctx_t ctx, CSTElement node) {
 
 wcl::doc Emitter::walk_literal(ctx_t ctx, CSTElement node) {
   MEMO(ctx, node);
-  MEMO_RET(walk_placeholder(ctx, node));
+  assert(node.id() == CST_LITERAL);
+
+  // clang-format off
+  auto multiline_str_fmt = fmt()
+    .match(
+      pred(TOKEN_MSTR_BEGIN, fmt().token(TOKEN_MSTR_BEGIN))
+     .pred(TOKEN_MSTR_RESUME, fmt().token(TOKEN_MSTR_RESUME))
+     // No otherwise, this should fail if neither are true
+    )
+    .fmt_while(
+      {TOKEN_NL, TOKEN_WS, TOKEN_MSTR_CONTINUE},
+      fmt().match(
+        pred(TOKEN_WS, fmt().token(TOKEN_WS))
+       .pred(TOKEN_NL, fmt().token(TOKEN_NL))
+       .pred(TOKEN_MSTR_CONTINUE, fmt().token(TOKEN_MSTR_CONTINUE))
+      ))
+    .match(
+      pred(TOKEN_MSTR_PAUSE, fmt().token(TOKEN_MSTR_PAUSE))
+     .pred(TOKEN_MSTR_END, fmt().token(TOKEN_MSTR_END))
+     // No otherwise, this should fail if neither are true
+    );
+  // clang-format on
+
+  auto node_fmt = fmt().walk(DISPATCH(walk_placeholder));
+  auto token_fmt = fmt().walk(WALK_TOKEN);
+
+  // clang-format off
+  MEMO_RET(fmt().match(
+    // TODO: starting 'pred()' function doesn't allow init lists
+    pred(ConstPredicate(false), fmt())
+   .pred({TOKEN_MSTR_BEGIN, TOKEN_MSTR_RESUME}, multiline_str_fmt)
+   .pred([](wcl::doc_builder&, ctx_t, CSTElement& node,
+                  const token_traits_map_t&){ return node.isNode(); }, node_fmt)
+   .otherwise(token_fmt))
+   .format(ctx, node.firstChildElement(), token_traits));
```

Comment:
What are all the literal cases that are being handled here? multi-line string and token are two...what are the others?

Also does this handle interpolation? It doesn't need to in this version I'm just curious if somehow it already does.

Reply:
> What are all the literal cases that are being handled here?
You got it, right now its just multi-line string and token. There are likely more cases but I'll circle back around to them when I hit examples of them. Right now I'm just trying to format increasingly more complicated wake source files.